pkg/controller/session/session_controller.go

```diff
@@ -70,80 +91,168 @@ func (r *ReconcileSession) Reconcile(request reconcile.Request) (reconcile.Resul
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	reqLogger.Info("Reconciling Session")
 
-	ctx := context.TODO()
+	ctx := model.SessionContext{
+		Context:   context.TODO(),
+		Name:      request.Name,
+		Namespace: request.Namespace,
+		Log:       reqLogger,
+		Client:    r.client,
+	}
 
 	// Fetch the Session instance
-	instance := &istiov1alpha1.Session{}
-	err := r.client.Get(ctx, request.NamespacedName, instance)
+	session := &istiov1alpha1.Session{}
+	err := r.client.Get(ctx, request.NamespacedName, session)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			// Request object not found, could have been deleted after reconcile request.
-			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
-			// Return and don't requeue
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
+	deleted := session.DeletionTimestamp != nil
+	if deleted {
+		reqLogger.Info("Deleted session")
+		if !session.HasFinalizer(finalizer) {
+			return reconcile.Result{}, nil
+		}
+	} else {
+		reqLogger.Info("Added session")
+		if !session.HasFinalizer(finalizer) {
+			session.AddFinalizer(finalizer)
+			if err := r.client.Update(ctx, session); err != nil {
+				ctx.Log.Error(err, "Failed to add finalizer on session")
+			}
+		}
+	}
 
-	reqLogger.Info("Added session", "name", request.Name, "namespace", request.Namespace)
+	refs := statusesToRef(*session)
+	if deleted {
+		deleteAllRefs(ctx, session, refs)
+	} else {
+		if err := syncAllRefs(ctx, session); err != nil {
+			return reconcile.Result{}, err
+		}
+		deleteRemovedRefs(ctx, session, refs)
+	}
 
-	deployment := appsv1.Deployment{}
+	if deleted {
+		session.RemoveFinalizer(finalizer)
+		if err := r.client.Update(ctx, session); err != nil {
+			ctx.Log.Error(err, "Failed to remove finalizer on session")
+		}
 
-	err = r.client.Get(ctx, types.NamespacedName{Namespace: request.Namespace, Name: instance.Spec.Ref}, &deployment)
-	if err != nil {
-		updateStatus(ctx, reqLogger, r.client, instance, fmt.Sprintf("%v", err))
-		return reconcile.Result{Requeue: false}, err
 	}
+	return reconcile.Result{}, nil
+}
 
-	reqLogger.Info("Found Deployment", "image", deployment.Spec.Template.Spec.Containers[0].Image)
-
-	serviceList := corev1.ServiceList{}
-	err = r.client.List(ctx, &client.ListOptions{Namespace: request.Namespace}, &serviceList)
-	if err != nil {
-		updateStatus(ctx, reqLogger, r.client, instance, fmt.Sprintf("%v", err))
-		return reconcile.Result{Requeue: false}, err
+func deleteRemovedRefs(ctx model.SessionContext, session *istiov1alpha1.Session, refs []*model.Ref) { //nolint[:hugeParam]
+	for _, ref := range refs {
+		found := false
+		for _, r := range session.Spec.Refs {
+			if ref.Name == r {
+				found = true
```

Comment:
Shouldn't we break out of the loop in such a case?

Reply:
https://github.com/aslakknutsen/istio-workspace/pull/32/commits/77ff4ddabe20874e8432cafd69400401b8ac047f